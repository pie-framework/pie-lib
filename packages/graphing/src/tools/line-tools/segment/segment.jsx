import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { types } from '@pie-lib/plot';

/**
 * A low level segment component
 */
export class RawSegment extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    correctness: PropTypes.string,
    disabled: PropTypes.bool,
    from: PropTypes.shape({
      x: PropTypes.number,
      y: PropTypes.number
    }).isRequired,
    to: PropTypes.shape({
      x: PropTypes.number,
      y: PropTypes.number
    }).isRequired,
    graphProps: types.GraphPropsType.isRequired
  };

  render() {
    const { classes, disabled, className, correctness, graphProps, from, to, ...rest } = this.props;
    const { scale } = graphProps;

    return (
      <line
        x1={scale.x(from.x)}
        y1={scale.y(from.y)}
        x2={scale.x(to.x)}
        y2={scale.y(to.y)}
        className={classNames(
          classes.bgSegment,
          disabled && classes.disabled,
          classes[correctness],
          className
        )}
        {...rest}
      />
    );
  }
}
