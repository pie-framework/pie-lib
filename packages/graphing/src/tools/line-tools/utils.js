import { calculateThirdPointOnLine } from '../../utils';

export const arrowDimensions = {
  none: 0,
  vector: 7,
  ray: 5,
  line: 5
};

export const calculateCorrectScaledPoints = (scale, from, to, type = 'none') => {
  const scaledFromX = scale.x(from.x);
  const scaledFromY = scale.y(from.y);
  let scaledToX = scale.x(to.x);
  let scaledToY = scale.y(to.y);
  let middlePoint;

  middlePoint = calculateThirdPointOnLine(
    { x: scaledToX, y: scaledToY },
    { x: scaledFromX, y: scaledFromY },
    {
      domain: {
        min: scale.x(to.x) - arrowDimensions[type],
        max: scale.x(to.x) + arrowDimensions[type]
      },
      range: {
        min: scale.y(to.y) - arrowDimensions[type],
        max: scale.y(to.y) + arrowDimensions[type]
      }
    }
  );

  if (middlePoint && isFinite(middlePoint.x) && isFinite(middlePoint.y)) {
    scaledToX = middlePoint.x;
    scaledToY = middlePoint.y;
  }

  return {
    x: scaledToX,
    y: scaledToY
  };
};
