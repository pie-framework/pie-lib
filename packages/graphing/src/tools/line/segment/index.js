import React from 'react';
import LineComponent from '../common/component';
import _ from 'lodash';

export const tool = () => ({
  type: 'segment',
  Component: LineComponent,
  addPoint: (point, mark) => {
    if (!mark) {
      return {
        type: 'segment',
        building: true,
        from: point
      };
    }

    if (_.isEqual(point, mark.from)) {
      return { ...mark };
    }

    return { ...mark, building: false, to: point };
  }
});
