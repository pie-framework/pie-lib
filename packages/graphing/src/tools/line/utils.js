import debug from 'debug';
// import { swap } from '../point/utils';

const log = debug('pie-lib:charting:line:utils');

const lineExpressionRegex = new RegExp(/([+|-]?\d+)?(x)?([+|-]\d+)?/i);

export class Expression {
  constructor(multiplier, b) {
    this.multiplier = multiplier;
    this.isVerticalLine = Math.abs(this.multiplier) === Infinity;
    this.b = this.isVerticalLine ? undefined : b;
  }

  expr() {
    return this.isVerticalLine ? 'x' : `${this.multiplier}x + ${this.b}`;
  }

  getY(x) {
    return this.isVerticalLine ? x : this.multiplier * x + this.b;
  }

  getX(y) {
    return this.isVerticalLine ? y : (y - this.b) / this.multiplier;
  }

  equals(other) {
    return this.multiplier === other.multiplier && this.b === other.b;
  }
}

export const pointsHaveSameExpression = (a, b) => {
  const expressionA = expression(a.from, a.to);
  const expressionB = expression(b.from, b.to);
  return expressionA.equals(expressionB);
};
/**
 * Create a linear expression from 2 points
 */

export const expression = (from, to) => {
  log('[expression] from: ', from, 'to: ', to);
  const multiplier = (to.y - from.y) / (to.x - from.x);
  const zeroedY = multiplier * from.x;
  const b = from.y - zeroedY;
  return new Expression(multiplier, b);
};

export const point = (x, y) => ({ x, y });

export const pointsFromExpression = expression => {
  const to = point(0, expression.b);
  const huh = expression.getX(0);
  const from = point(huh, 0);

  return {
    from,
    to
  };
};

export const expressionFromDescriptor = descriptor => {
  const lineDescriptor = descriptor.match(lineExpressionRegex);
  const maybeB = parseInt(lineDescriptor[3], 10);
  const maybeMultiplier = parseInt(lineDescriptor[1], 10);
  let multiplier = isNaN(maybeMultiplier)
    ? lineDescriptor[2] === undefined
      ? 0
      : 1
    : maybeMultiplier;
  let b = isNaN(maybeB) ? 0 : maybeB;

  // it's a constant, no variable found
  if (lineDescriptor[2] === undefined) {
    b = multiplier;
    multiplier = 0;
  }

  return new Expression(multiplier, b);
};

export const hasLine = (lines, line) => lineIndex(lines, line) !== -1;

export const removeLine = (lines, line) => {
  const index = lineIndex(lines, line);
  if (index === -1) {
    return lines;
  }
  const out = lines.slice();
  out.splice(index, 1);
  return out;
};

export const removeLines = (lines, toRemove) => {
  log('[removeLines] points: ', lines, toRemove);
  toRemove = Array.isArray(toRemove) ? toRemove : [toRemove];
  const out = lines.filter(p => !hasLine(toRemove, p));
  log('[removePoints] points: ', out);
  return out;
};

export const pointsEqual = (a, b) => a.x == b.x && a.y == b.y;

export const linesEqual = (a, b) => pointsEqual(a.from, b.from) && pointsEqual(a.to, b.to);

export const lineIndex = (lines, line) => lines.findIndex(l => linesEqual(l, line));

// export const swapLine = (lines, from, to) => swap(lines, from, to, lineIndex);
