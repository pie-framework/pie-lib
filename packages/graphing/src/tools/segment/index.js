import Segment from './component';
import _ from 'lodash';

export const tool = () => ({
  type: 'segment',
  Component: Segment,
  addPoint: (point, mark) => {
    if (!mark) {
      return {
        type: 'segment',
        building: true,
        firstEnd: point
      };
    }

    if (_.isEqual(point, mark.firstEnd)) {
      return { ...mark };
    }

    return { ...mark, building: false, secondEnd: point };
  }
});
