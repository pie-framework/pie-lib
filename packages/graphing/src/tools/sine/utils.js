import _ from 'lodash';
import debug from 'debug';
import Point from '@mapbox/point-geometry';
const log = debug('pie-lib:graphing;sine:utils');

/**
 * build the x points with root as the anchor.
 * adds padding left and right so that
 * the anchor can be moved left/right
 * and there will be points off stage rendering.
 *
 *  root: 0, freq: 2, min: -5, max: 5 ==> -5,-4-2,0,2,4,5
 */
export const xPoints = (root, freq, min, max) => {
  return _.concat(
    min - freq,
    _.rangeRight(root, min - freq, freq * -1),
    _.range(root + freq, max + freq, freq),
    max + freq
  );
};

export const sinY = (amplitude, freq) => x => {
  const num = 2 * Math.PI * x;
  const frac = num / freq;
  return amplitude * parseFloat(Math.sin(frac).toFixed(10));
};

export const buildDataPoints = (min, max, root, interval, yFn) => {
  log('[buildDataPoints] min:', min, 'max:', max, 'root:', root);
  // const fn = sinY(amplitude, freq);
  const diff = max - root.x;
  const plusDiff = min - root.x;
  const xs = xPoints(0, interval, min - root.x - diff, max - root.x - plusDiff);
  log('xs:', xs);
  return xs.map(v => new Point(v, yFn(v))).map(p => p.add(new Point(root.x, root.y)));
};
