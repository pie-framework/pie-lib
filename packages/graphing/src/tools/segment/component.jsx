import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { ToolPropType } from '../types';
import debug from 'debug';
import BasePoint from '../point/base-point';
import Segment from './segment';
import { point } from '../../utils';
import classNames from 'classnames';
import { types } from '@pie-lib/plot';
import isEqual from 'lodash/isEqual';

const log = debug('pie-lib:graphing:segment');

const opacityPulsate = opacity => ({
  '0%': { opacity: '0.0' },
  '50%': { opacity },
  '100%': { opacity: '0.0' }
});

export const PointType = {
  x: PropTypes.number,
  y: PropTypes.number
};

export class RawBaseSegment extends React.Component {
  static propTypes = {
    building: PropTypes.bool,
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    correctness: PropTypes.string,
    firstEnd: PropTypes.shape(PointType),
    secondEnd: PropTypes.shape(PointType),
    disabled: PropTypes.bool,
    onChange: PropTypes.func.isRequired,
    onDragStart: PropTypes.func,
    onDragStop: PropTypes.func,
    graphProps: types.GraphPropsType.isRequired
  };

  static defaultProps = {};

  constructor(props) {
    super(props);
    this.state = {
      draggedFirstEnd: undefined
    };
  }

  moveFirstEnd = firstEnd => {
    const { onChange, secondEnd } = this.props;
    const d = { firstEnd, secondEnd };

    this.setState({ draggedFirstEnd: undefined }, () => {
      if (!isEqual(firstEnd, secondEnd)) {
        onChange(d);
      }
    });
  };

  dragFirstEnd = draggedFirstEnd => {
    log('[dragFirstEnd] ', draggedFirstEnd);
    this.setState({ draggedFirstEnd });
  };

  moveSecondEnd = secondEnd => {
    const { onChange, firstEnd } = this.props;
    const d = { firstEnd, secondEnd };
    this.setState({ draggedSecondEnd: undefined }, () => {
      if (!isEqual(secondEnd, firstEnd)) {
        onChange(d);
      }
    });
  };

  dragSecondEnd = draggedSecondEnd => {
    this.setState({ draggedSecondEnd });
  };

  dragSegment = draggedFirstEnd => {
    const { firstEnd, secondEnd } = this.props;

    log('dragSegment: ', draggedFirstEnd, 'firstEnd: ', firstEnd, 'secondEnd: ', secondEnd);

    const diff = point(firstEnd).sub(point(draggedFirstEnd));
    const draggedSecondEnd = point(secondEnd).sub(diff);

    this.setState({ draggedFirstEnd, draggedSecondEnd, isSegmentDrag: true });
  };

  moveSegment = firstEnd => {
    const { onChange, firstEnd: propsFirstEnd, secondEnd } = this.props;
    const diff = point(propsFirstEnd).sub(point(firstEnd));
    const d = {
      firstEnd,
      secondEnd: point(secondEnd).sub(diff)
    };
    this.setState(
      {
        draggedFirstEnd: undefined,
        draggedSecondEnd: undefined,
        isSegmentDrag: false
      },
      () => onChange(d)
    );
  };

  render() {
    let {
      firstEnd,
      disabled,
      classes,
      building,
      onDragStart,
      onDragStop,
      correctness,
      graphProps,
      secondEnd
    } = this.props;
    const { draggedFirstEnd, draggedSecondEnd, isSegmentDrag } = this.state;
    log('[render] draggedFirstEnd: ', draggedFirstEnd, 'firstEnd: ', firstEnd);

    secondEnd = secondEnd || firstEnd;
    const f = draggedFirstEnd || firstEnd;
    const s = draggedSecondEnd || secondEnd;
    const common = { graphProps };

    return (
      <g>
        <Segment
          disabled={building || disabled}
          correctness={correctness}
          className={classNames(building && classes.segmentBuilding)}
          x={isSegmentDrag ? firstEnd.x : f.x}
          y={isSegmentDrag ? firstEnd.y : f.y}
          firstEnd={isSegmentDrag ? firstEnd : f}
          secondEnd={isSegmentDrag ? secondEnd : s}
          onDrag={this.dragSegment}
          onMove={this.moveSegment}
          onDragStart={onDragStart}
          onDragStop={onDragStop}
          {...common}
        />
        <BasePoint
          disabled={building || disabled}
          correctness={correctness}
          x={isSegmentDrag ? s.x : secondEnd.x}
          y={isSegmentDrag ? s.y : secondEnd.y}
          onMove={this.moveSecondEnd}
          onDrag={this.dragSecondEnd}
          onDragStart={onDragStart}
          onDragStop={onDragStop}
          {...common}
        />
        <BasePoint
          disabled={building || disabled}
          correctness={correctness}
          x={isSegmentDrag ? f.x : firstEnd.x}
          y={isSegmentDrag ? f.y : firstEnd.y}
          onMove={this.moveFirstEnd}
          onDrag={this.dragFirstEnd}
          onDragStart={onDragStart}
          onDragStop={onDragStop}
          {...common}
        />
      </g>
    );
  }
}

export const BaseSegment = withStyles(theme => ({
  outerLine: {
    fill: 'rgb(0,0,0,0)',
    stroke: theme.palette.primary.light,
    strokeWidth: 4,
    '&:hover': {
      strokeWidth: 6,
      stroke: theme.palette.primary.dark
    }
  },
  segmentBuilding: {
    stroke: theme.palette.secondary.light,
    animation: 'opacityPulse 2s ease-out',
    animationIterationCount: 'infinite',
    opacity: 1
  },
  '@keyframes opacityPulse': opacityPulsate('0.3')
}))(RawBaseSegment);

export default class Component extends React.Component {
  static propTypes = {
    ...ToolPropType,
    graphProps: types.GraphPropsType.isRequired
  };

  change = d => {
    const { onChange } = this.props;
    const m = {
      ...this.props.mark,
      firstEnd: d.firstEnd,
      secondEnd: d.secondEnd
    };
    onChange(this.props.mark, m);
  };

  render() {
    const { mark, onDragStart, onDragStop, graphProps } = this.props;
    return (
      <BaseSegment
        {...mark}
        onChange={this.change}
        onDragStart={onDragStart}
        onDragStop={onDragStop}
        graphProps={graphProps}
      />
    );
  }
}
