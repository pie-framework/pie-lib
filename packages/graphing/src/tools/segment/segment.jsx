import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import { types, gridDraggable } from '@pie-lib/plot';
import * as utils from '../../utils';
import { disabled, correct, incorrect } from '../styles';

/**
 * A low level segment component
 *
 * TODO: This and base point have a lot of similarities - merge commonality
 *
 */
class RawSegment extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    correctness: PropTypes.string,
    disabled: PropTypes.bool,
    firstEnd: PropTypes.shape({
      x: PropTypes.number,
      y: PropTypes.number
    }).isRequired,
    secondEnd: PropTypes.shape({
      x: PropTypes.number,
      y: PropTypes.number
    }).isRequired,
    graphProps: types.GraphPropsType.isRequired
  };

  render() {
    const {
      classes,
      disabled,
      className,
      correctness,
      graphProps,
      firstEnd,
      secondEnd,
      ...rest
    } = this.props;
    const { scale } = graphProps;

    return (
      <line
        x1={scale.x(firstEnd.x)}
        y1={scale.y(firstEnd.y)}
        x2={scale.x(secondEnd.x)}
        y2={scale.y(secondEnd.y)}
        className={classNames(
          classes.bgSegment,
          disabled && classes.disabled,
          classes[correctness],
          className
        )}
        {...rest}
      />
    );
  }
}

const applyStyle = fn => ({
  ...fn('stroke'),
  '&:hover': {
    strokeWidth: 3,
    ...fn('stroke')
  }
});

const styles = theme => ({
  bgSegment: {
    fill: 'transparent',
    stroke: theme.palette.primary.light,
    strokeWidth: 3,
    transition: 'stroke 200ms ease-in, stroke-width 200ms ease-in',
    '&:hover': {
      strokeWidth: 6,
      stroke: theme.palette.primary.dark
    }
  },
  disabled: applyStyle(disabled),
  correct: applyStyle(correct),
  incorrect: applyStyle(incorrect)
});

export const BgSegment = withStyles(styles)(RawSegment);

export default gridDraggable({
  bounds: (props, { domain, range }) => {
    const { firstEnd, secondEnd } = props;
    const area = utils.lineToArea(firstEnd, secondEnd);

    return utils.bounds(area, domain, range);
  },
  anchorPoint: props => {
    const { firstEnd } = props;

    return firstEnd;
  },
  fromDelta: (props, delta) => {
    return utils.point(props).add(utils.point(delta));
  }
})(BgSegment);
