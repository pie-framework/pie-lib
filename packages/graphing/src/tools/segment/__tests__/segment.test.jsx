import { shallow } from 'enzyme';
import React from 'react';
import { BgSegment } from '../segment';
import { graphProps } from '../../../__tests__/utils';

describe('BgCircle', () => {
  let w;
  let onChange = jest.fn();
  const wrapper = extras => {
    const defaults = {
      classes: {},
      className: 'className',
      onChange,
      firstEnd: {
        x: 0,
        y: 0
      },
      secondEnd: {
        x: 0,
        y: 0
      },
      graphProps: graphProps()
    };
    const props = { ...defaults, ...extras };
    return shallow(<BgSegment {...props} />);
  };
  describe('snapshot', () => {
    it('renders', () => {
      w = wrapper();
      expect(w).toMatchSnapshot();
    });
  });
  describe('logic', () => {});
});
