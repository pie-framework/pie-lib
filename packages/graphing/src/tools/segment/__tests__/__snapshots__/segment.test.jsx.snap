// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BgCircle snapshot renders 1`] = `
<RawSegment
  className="className"
  classes={
    Object {
      "bgSegment": "RawSegment-bgSegment-1",
      "correct": "RawSegment-correct-3",
      "disabled": "RawSegment-disabled-2",
      "incorrect": "RawSegment-incorrect-4",
    }
  }
  firstEnd={
    Object {
      "x": 0,
      "y": 0,
    }
  }
  graphProps={
    Object {
      "domain": Object {
        "max": 1,
        "min": 0,
        "step": 1,
      },
      "range": Object {
        "max": 1,
        "min": 0,
        "step": 1,
      },
      "scale": Object {
        "x": [MockFunction],
        "y": [MockFunction],
      },
      "size": Object {
        "height": 400,
        "width": 400,
      },
      "snap": Object {
        "x": [MockFunction],
        "y": [MockFunction],
      },
    }
  }
  onChange={[MockFunction]}
  secondEnd={
    Object {
      "x": 0,
      "y": 0,
    }
  }
/>
`;
